### PR TITLE
IndexMetaData: Introduce internal format index setting

### PR DESCRIPTION
**Pull request by spinscale**
This setting is supposed to ease index upgrades as it allows you
to check for a new setting called `index.internal.version` which
in turn can be configured in the index settings and has to be set
to '6' in order to be valid.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 1]
```diff
@@ -259,6 +258,14 @@ public static State fromString(String state) {
                       Setting.Property.Dynamic,
                       Setting.Property.IndexScope);
 
+    /**
+     * an internal index format description, allowing us to find out if this index is upgraded or needs upgrading
+     */
+    private static final String INDEX_INTERNAL_FORMAT = "index.internal.format";
+    public static final Setting<Integer> INDEX_INTERNAL_FORMAT_SETTING =
+            Setting.intSetting(INDEX_INTERNAL_FORMAT, 0, Setting.Property.IndexScope);
+    static final int INTERNAL_INDEX_FORMAT_CURRENT = 6;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: maybe we make this a bit more generic and call it `index.format` and `INTERNAL_INDEX_FORMAT_CURRENT` should not be here. it's up to every user of this setting to define the current?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 2]
```diff
@@ -279,6 +286,7 @@ public static State fromString(String state) {
     private final Index index;
     private final long version;
     private final long[] primaryTerms;
+    private final int internalIndexFormat;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: do we really have to make this a first class citizen or can we just leave it as a setting?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 3]
```diff
@@ -259,6 +258,14 @@ public static State fromString(String state) {
                       Setting.Property.Dynamic,
                       Setting.Property.IndexScope);
 
+    /**
+     * an internal index format description, allowing us to find out if this index is upgraded or needs upgrading
+     */
+    private static final String INDEX_INTERNAL_FORMAT = "index.internal.format";
+    public static final Setting<Integer> INDEX_INTERNAL_FORMAT_SETTING =
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: also make this setting `Setting.Property.Final`

**Review by s1monw — CHANGES_REQUESTED** [review]
left some comments

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java (hunk 1 above)
Comment (spinscale) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: fixed

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java (hunk 2 above)
Comment (spinscale) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: makes more sense as a setting. fixed

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java (hunk 3 above)
Comment (spinscale) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: oooh, wasnt aware of that. great!

**Comment by spinscale** [comment]
all comments made a ton of sense. incorporated.

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 4]
```diff
@@ -309,7 +315,7 @@ private IndexMetaData(Index index, long version, long[] primaryTerms, State stat
                           ImmutableOpenMap<String, Custom> customs, ImmutableOpenIntMap<Set<String>> inSyncAllocationIds,
                           DiscoveryNodeFilters requireFilters, DiscoveryNodeFilters initialRecoveryFilters, DiscoveryNodeFilters includeFilters, DiscoveryNodeFilters excludeFilters,
                           Version indexCreatedVersion, Version indexUpgradedVersion,
-                          int routingNumShards, int routingPartitionSize, ActiveShardCount waitForActiveShards) {
+                          int routingNumShards, int routingPartitionSize, ActiveShardCount waitForActiveShards, int internalIndexFormat) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: this is unused?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 5]
```diff
@@ -1051,9 +1057,11 @@ public IndexMetaData build() {
             }
 
             final String uuid = settings.get(SETTING_INDEX_UUID, INDEX_UUID_NA_VALUE);
+
+            int internalIndexFormat = INDEX_FORMAT_SETTING.get(settings);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: this is unneeded see above

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 6]
```diff
@@ -1051,9 +1057,11 @@ public IndexMetaData build() {
...
             return new IndexMetaData(new Index(index, uuid), version, primaryTerms, state, numberOfShards, numberOfReplicas, tmpSettings, mappings.build(),
                 tmpAliases.build(), customs.build(), filledInSyncAllocationIds.build(), requireFilters, initialRecoveryFilters, includeFilters, excludeFilters,
-                indexCreatedVersion, indexUpgradedVersion, getRoutingNumShards(), routingPartitionSize, waitForActiveShards);
+                indexCreatedVersion, indexUpgradedVersion, getRoutingNumShards(), routingPartitionSize, waitForActiveShards, internalIndexFormat);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: this should go away

**Review by s1monw — CHANGES_REQUESTED** [review]
left some comments

**Review by s1monw — APPROVED** [review]
LGTM

**Pull request merged**